warp10/src/main/java/io/warp10/script/formatted/FormattedWarpScriptFunction.java

```diff
@@ -0,0 +1,305 @@
+        if (!map.containsKey(arg.getName())) {
+
+          throw new WarpScriptException("The MAP that is on top of the stack does not have the argument '" + arg.getName() +
+            "' (of type "  + arg.WarpScriptType() + ") that is required by " + getName());
+        }
+
+        if (!arg.getClazz().isInstance(map.get(arg.getName()))) {
+
+          throw new WarpScriptException(getClass().getCanonicalName() + " expects the argument '" + arg.getName() + "' to" +
+            " be a " + arg.WarpScriptType() + ".");
+        }
+      }
+
+      //
+      // Check that optional args that are in the map are of the correct type
+      //
+
+      for (ArgumentSpecification arg: optArgs) {
+        if (map.containsKey(arg.getName()) && !arg.getClazz().isInstance(map.get(arg.getName()))) {
+
+          throw new WarpScriptException(getClass().getCanonicalName() + " expects the argument '" + arg.getName() + "' to " +
+            "be a " + arg.WarpScriptType() + ".");
+        }
+      }
+
+      //
+      // Consume the top of the stack
+      //
+
+      formattedArgs = new HashMap<String, Object> ((Map) stack.pop());
+
+    } else {
+
+      //
+      // Case 2: No optional argument are given
+      //
+
+      if (stack.depth() < args.size()) {
+        throw new WarpScriptException(getClass().getCanonicalName() + " expects to find " + args.size() + " arguments" +
+          " off the top of the stack, but the stack contains only " + stack.depth() + " levels.");
+      }
+
+      //
+      // Check argument types
+      //
+
+      for (int i = 0; i < args.size(); i++) {
+        ArgumentSpecification arg = args.get(args.size() - 1 - i);
+        Object candidate = stack.get(i);
+
+        if (!arg.getClazz().isInstance(candidate)) {
+          throw new WarpScriptException(getClass().getCanonicalName() + " expects to find a '" + arg.getName() + "' (a " +
+            arg.WarpScriptType() + ") " + leveldenomination(i));
+        }
+      }
+
+      //
+      // Consume these arguments off the top of the stack
+      //
+
+      formattedArgs = new HashMap<String,Object>();
+
+      for (int i = 0; i < args.size(); i++) {
+        ArgumentSpecification arg = args.get(args.size() - 1 - i);
+        formattedArgs.put(arg.getName(), stack.pop());
+      }
+
+    }
+
+    //
+    // Set absent optional arguments to their default values
+    //
+
+    for (ArgumentSpecification arg: optArgs) {
+      if (!formattedArgs.containsKey(arg.getName())) {
+        formattedArgs.put(arg.getName(), arg.getDefaultValue());
+      }
+    }
+
+    return apply(formattedArgs, stack);
+  }
+
+  final private String leveldenomination(int i) {
+    if (i < 0) {
+      throw new IllegalStateException("Can not be negative");
+    }
+
+    if (0 == i) {
+      return "on top of the stack.";
+    } else if (1 == i) {
+      return "below the top of the stack.";
+    } else if (2 == i) {
+      return "on 3rd position counting from the top of the stack.";
```

Comment:
We call those positions 'levels',  the top of the stack is also level 1, so if i == 2, you are at level 3, please use this terminology, i.e. stack level #